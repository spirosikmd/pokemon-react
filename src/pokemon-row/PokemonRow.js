import React, { PureComponent } from 'react';

const styles = {
  myPokemon: {
    color: 'red',
  },
};

class PokemonRow extends PureComponent {
  handlePokemonClick(pokemonName) {
    this.props.onPokemonClick(pokemonName);
  }

  handleCatchClick(pokemonName) {
    this.props.onCatchClick(pokemonName);
  }

  getStyle(isMyPokemon) {
    return isMyPokemon ? Object.assign({}, styles.myPokemon) : {};
  }

  render() {
    const pokemon = this.props.pokemon;

    return (
      <li>
        <span
          onClick={() => this.handlePokemonClick(pokemon.name)}
          style={this.getStyle(this.props.isMyPokemon)}
        >
          {pokemon.name}
        </span>
        <button onClick={() => this.handleCatchClick(pokemon.name)}>
          {this.props.isMyPokemon ? 'set free..' : 'catch!'}
        </button>
      </li>
    );
  }
}

export default PokemonRow;
