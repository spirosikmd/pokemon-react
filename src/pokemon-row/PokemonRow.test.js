import React from 'react';
import renderer from 'react-test-renderer';
import PokemonRow from './PokemonRow';

const pokemon = {
  name: 'pikachu',
};

const mockHandleOnPokemonClick = jest.fn();
const mockHandleOnFavoriteClick = jest.fn();

test('PokemonRow renders pokemon name favorite button and handle data change', () => {
  const component = renderer.create(
    <PokemonRow
      pokemon={pokemon}
      onPokemonClick={mockHandleOnPokemonClick}
      onFavoriteClick={mockHandleOnFavoriteClick}
    />
  );
  let tree = component.toJSON();
  expect(tree).toMatchSnapshot();

  const name = tree.children[0];
  name.props.onClick();
  expect(mockHandleOnPokemonClick.mock.calls.length).toBe(1);
  expect(mockHandleOnPokemonClick.mock.calls[0][0]).toBe('pikachu');

  const favorite = tree.children[1];
  favorite.props.onClick();
  expect(mockHandleOnFavoriteClick.mock.calls.length).toBe(1);
  expect(mockHandleOnFavoriteClick.mock.calls[0][0]).toBe('pikachu');
});

test('PokemonRow renders a "*" when pokemon is not favorite and style color red', () => {
  const component = renderer.create(
    <PokemonRow pokemon={pokemon} isFavorite={false} />
  );
  let tree = component.toJSON();
  expect(tree).toMatchSnapshot();
});

test('PokemonRow renders a "-" when pokemon is favorite', () => {
  const component = renderer.create(
    <PokemonRow pokemon={pokemon} isFavorite={true} />
  );
  let tree = component.toJSON();
  expect(tree).toMatchSnapshot();
});

test('PokemonRow getStyle returns empty object when isFavorite is false', () => {
  const pokemonRow = new PokemonRow();
  expect(pokemonRow.getStyle(false)).toEqual({});
});

test('PokemonRow getStyle returns style with color red when isFavorite is true', () => {
  const pokemonRow = new PokemonRow();
  expect(pokemonRow.getStyle(true)).toEqual({
    color: 'red',
  });
});
