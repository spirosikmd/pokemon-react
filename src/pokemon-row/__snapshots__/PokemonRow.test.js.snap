exports[`test PokemonRow renders a "catch!" when pokemon is not my pokemon and style color red 1`] = `
<li>
  <span
    onClick={[Function]}
    style={Object {}}>
    pikachu
  </span>
  <button
    onClick={[Function]}>
    catch!
  </button>
</li>
`;

exports[`test PokemonRow renders a "set free.." when pokemon is my pokemon 1`] = `
<li>
  <span
    onClick={[Function]}
    style={
      Object {
        "color": "red",
      }
    }>
    pikachu
  </span>
  <button
    onClick={[Function]}>
    set free..
  </button>
</li>
`;

exports[`test PokemonRow renders pokemon name catch button and handle data change 1`] = `
<li>
  <span
    onClick={[Function]}
    style={Object {}}>
    pikachu
  </span>
  <button
    onClick={[Function]}>
    catch!
  </button>
</li>
`;
